File: app/src/main/java/igrek/todotree/domain/treeitem/LinkTreeItem.java
package igrek.todotree.domain.treeitem;


import com.google.common.base.Joiner;

import java.util.List;

import igrek.todotree.commands.TreeCommand;

public class LinkTreeItem extends AbstractTreeItem {
	
	private String targetPath;
	private String customName;
	
	public LinkTreeItem(AbstractTreeItem parent, String targetPath, String customName) {
		super(parent);
		this.targetPath = targetPath;
		this.customName = customName;
	}
	
	@Override
	public LinkTreeItem clone() {
		return new LinkTreeItem(null, targetPath, customName).copyChildren(this);
	}
	
	@Override
	public String getDisplayName() {
		if (hasCustomName())
			return customName;
		
		AbstractTreeItem target = getTarget();
		if (target == null) {
			return getDisplayTargetPath();
		} else {
			return target.getDisplayName();
		}
	}
	
	@Override
	public String getTypeName() {
		return "link";
	}
	
	public void setCustomName(String customName) {
		this.customName = customName;
	}
	
	public String getCustomName() {
		return customName;
	}
	
	public boolean hasCustomName() {
		return customName != null;
	}
	
	public String getTargetPath() {
		return targetPath;
	}
	
	public String getDisplayTargetPath() {
		return "/" + targetPath.replace("\t", "/");
	}
	
	public AbstractTreeItem getTarget() {
		String[] paths = targetPath.split("\\t");
		return new TreeCommand().findItemByPath(paths);
	}
	
	public void setTarget(AbstractTreeItem target) {
		Joiner joiner = Joiner.on("\t");
		List<String> names = target.getNamesPaths();
		targetPath = joiner.join(names);
	}
	
	public void setTarget(AbstractTreeItem targetParent, String targetName) {
		Joiner joiner = Joiner.on("\t");
		List<String> names = targetParent.getNamesPaths();
		names.add(targetName);
		targetPath = joiner.join(names);
	}
	
	public boolean isBroken() {
		return getTarget() == null;
	}
	
}

File: app/src/main/java/igrek/todotree/domain/treeitem/LinkTreeItem.kt
package igrek.todotree.domain.treeitem

import com.google.common.base.Joiner
import igrek.todotree.intent.TreeCommand

class LinkTreeItem(
        parent: AbstractTreeItem?,
        var targetPath: String,
        var customName: String?,
) : AbstractTreeItem(parent) {

    override fun clone(): LinkTreeItem {
        return LinkTreeItem(null, targetPath, customName).copyChildren(this)
    }

    override val displayName: String
        get() {
            if (hasCustomName())
                return customName.orEmpty()
            val target = target
            return target?.displayName ?: displayTargetPath
        }

    override val typeName: String
        get() = "link"

    fun hasCustomName(): Boolean {
        return customName != null
    }

    val displayTargetPath: String
        get() = "/" + targetPath.replace("\t", "/")

    val target: AbstractTreeItem?
        get() {
            val paths = targetPath.split("\\t").toTypedArray()
            return TreeCommand().findItemByPath(paths)
        }

    fun setTarget(target: AbstractTreeItem) {
        val joiner = Joiner.on("\t")
        val names = target.namesPaths
        targetPath = joiner.join(names)
    }

    fun setTarget(targetParent: AbstractTreeItem, targetName: String?) {
        val joiner = Joiner.on("\t")
        val names = targetParent.namesPaths
        names.add(targetName)
        targetPath = joiner.join(names)
    }

    val isBroken: Boolean
        get() = target == null

}